https/tls_config_test.go

```diff
@@ -0,0 +1,361 @@
+		"YAML error":                   regexp.MustCompile(`yaml`),
+		"Invalid ClientAuth":           regexp.MustCompile(`ClientAuth`),
+		"TLS handshake":                regexp.MustCompile(`tls`),
+		"HTTP Request to HTTPS server": regexp.MustCompile(`HTTP`),
+	}
+)
+
+func getPort() string {
+	listener, err := net.Listen("tcp", ":0")
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer listener.Close()
+	p := listener.Addr().(*net.TCPAddr).Port
+	return fmt.Sprintf(":%v", p)
+}
+
+type TestInputs struct {
+	Name           string
+	Server         func() *http.Server
+	UseNilServer   bool
+	YAMLConfigPath string
+	ExpectedError  *regexp.Regexp
+	UseTLSClient   bool
+}
+
+func TestYAMLFiles(t *testing.T) {
+	testTables := []*TestInputs{
+		{
+			Name:           `path to config yml invalid`,
+			YAMLConfigPath: "somefile",
+			ExpectedError:  ErrorMap["No such file"],
+		},
+		{
+			Name:           `empty config yml`,
+			YAMLConfigPath: "testdata/tls_config_empty.yml",
+			ExpectedError:  ErrorMap["No such file"],
+		},
+		{
+			Name:           `invalid config yml (invalid structure)`,
+			YAMLConfigPath: "testdata/tls_config_junk.yml",
+			ExpectedError:  ErrorMap["YAML error"],
+		},
+		{
+			Name:           `invalid config yml (cert path empty)`,
+			YAMLConfigPath: "testdata/tls_config_noAuth_certPath_empty.bad.yml",
+			ExpectedError:  ErrorMap["No such file"],
+		},
+		{
+			Name:           `invalid config yml (key path empty)`,
+			YAMLConfigPath: "testdata/tls_config_noAuth_keyPath_empty.bad.yml",
+			ExpectedError:  ErrorMap["No such file"],
+		},
+		{
+			Name:           `invalid config yml (cert path and key path empty)`,
+			YAMLConfigPath: "testdata/tls_config_noAuth_certPath_keyPath_empty.bad.yml",
+			ExpectedError:  ErrorMap["No such file"],
+		},
+		{
+			Name:           `invalid config yml (cert path invalid)`,
+			YAMLConfigPath: "testdata/tls_config_noAuth_certPath_invalid.bad.yml",
+			ExpectedError:  ErrorMap["No such file"],
+		},
+		{
+			Name:           `invalid config yml (key path invalid)`,
+			YAMLConfigPath: "testdata/tls_config_noAuth_keyPath_invalid.bad.yml",
+			ExpectedError:  ErrorMap["No such file"],
+		},
+		{
+			Name:           `invalid config yml (cert path and key path invalid)`,
+			YAMLConfigPath: "testdata/tls_config_noAuth_certPath_keyPath_invalid.bad.yml",
+			ExpectedError:  ErrorMap["No such file"],
+		},
+		{
+			Name:           `invalid config yml (invalid ClientAuth)`,
+			YAMLConfigPath: "testdata/tls_config_noAuth.bad.yml",
+			ExpectedError:  ErrorMap["Invalid ClientAuth"],
+		},
+		{
+			Name:           `invalid config yml (invalid ClientCAs filepath)`,
+			YAMLConfigPath: "testdata/tls_config_auth_clientCAs_invalid.bad.yml",
+			ExpectedError:  ErrorMap["No such file"],
+		},
+	}
+	for _, testInputs := range testTables {
+		t.Run(testInputs.Name, testInputs.Test)
+	}
+}
+
+func TestServerBehaviour(t *testing.T) {
+	testTables := []*TestInputs{
+		{
+			Name:           `nil Server and default client`,
```

Comment:
By definition, the `Listen()` function expects that `server` isn't `nil` so I wouldn't have this test.